Add footnote warning to hashing a hash

FWDekker opened the pull request:
It's generally not recommended to use bcrypt on top of another hashing algorithm. While doing it over sha256 is obviously not as bad as doing it over md5, I think it's important to note in the README that OWASP recommends against the practice.

charmander reviewed (changes requested):
> While doing it over sha256 is obviously not as bad as doing it over md5

Sure it is. “Hash shucking” remains an issue, and the differences in collision resistance and output size aren’t relevant to this application.

How about recommending pre-hashing with HMAC as a keyed hash instead? Or just `bcrypt.kdf`?

FWDekker commented on README.rst:
```diff
@@ -231,6 +231,8 @@ encode it to prevent NULL byte problems before hashing the result with
     ...     bcrypt.gensalt()
     ... )
 
+Note, however, that this practice is generally `recommended against`_, as it may expose the system to  `hash shucking`_ attacks, and to denial of service attacks if an attacker rapidly sends many extremely long passwords.
```
You're right. Even a 1GB password is only a factor 2 slower than 12-round bcrypt on my machine, and my server already refuses a request if its body is larger than 16MB. I'll remove it from the README.

FWDekker commented:
> Sure it is. “Hash shucking” remains an issue, and the differences in collision resistance and output size aren’t relevant to this application.

You're absolutely right!

> How about recommending pre-hashing with HMAC as a keyed hash instead? Or just `bcrypt.kdf`?

I've removed the mention of the DDoS attack, replaced `sha256` with `bcrypt_pbkdf`, and adjusted the explanation of why using an unkeyed hash function is not a good idea. What do you think?

I don't have a lot of experience with `bcrypt_pbkdf`, so I just copied the parameters from the kdf part of the README. Do you think those parameters are readable?

reaperhulk commented:
https://flak.tedunangst.com/post/new-openssh-key-format-and-bcrypt-pbkdf has some information about choosing parameters for bcrypt_pbkdf. I've fixed CircleCI so if you push a new commit here CI should pass now and we can get it reviewed 😄

charmander commented on README.rst:
```diff
@@ -217,20 +217,29 @@ As of 3.0.0 the ``$2y$`` prefix is still supported in ``hashpw`` but deprecated.
     >>> password = b"an incredibly long password" * 10
     >>> hashed = bcrypt.hashpw(
-    ...     base64.b64encode(hashlib.sha256(password).digest()),
+    ...     base64.b64encode(bcrypt.kdf(password=password,
```
Ah, I meant using `bcrypt.kdf` directly instead of `bcrypt.hashpw`. This works, but it… feels unusual? If people want to keep the descriptive hash format that `hashpw` creates, HMAC seems simpler (although I guess the only *concrete* difference is that you don’t have to make an arbitrary choice for the number of output bytes and rounds).

FWDekker replied:
I see what you mean. I just literally placed one in the other, but there's more intuitive ways of doing this.

Do you mean `base64.b64encode(bcrypt.kdf(password=password, salt=bcrypt.gensalt(), <whatever>))`? Would password verification simply be to repeat the procedure and compare strings, or is there an analogue to `hashpw`?

And with the HMAC one, do you mean `bcrypt.hashpw(base64.b64encode(hmac.digest(pepper, password, "sha256")), bcrypt.gensalt())`?

If so, I suppose another important difference is that the latter requires storing a pepper (or per-hash salt) separately.

FWDekker replied:
I had forgotten about this PR, but it's now almost a year old.

Do you perhaps have time to take a look at my question to see if I understood your suggestion correctly?

charmander replied:
> Do you mean `base64.b64encode(bcrypt.kdf(password=password, salt=bcrypt.gensalt(), <whatever>))`? Would password verification simply be to repeat the procedure and compare strings, or is there an analogue to `hashpw`?
> 
> And with the HMAC one, do you mean `bcrypt.hashpw(base64.b64encode(hmac.digest(pepper, password, "sha256")), bcrypt.gensalt())`?

Yes and yes.

On second thought, though… maybe the best answer is replacing the code with your original footnote, and pointing back to “but you should really use argon2id” at the top of the README?

FWDekker replied:
lol, good point. I've made sure to emphasise that not using bcrypt is probably the best solution. However, I'm sure there will be people who will somehow be forced to use bcrypt, either because of legacy software, weird interoperability, legal requirements, incompetent management, you name it. So I think it is still worthwhile to explain how to work around the length limitation. I've also chosen to use the HMAC variant because if a reader doesn't already know about hash shucking and peppers, I think it's unlikely they'll manage to choose reasonable numbers of rounds and bytes.

Let me know what you think of the rewritten section.

reaperhulk commented:
closing/reopening to trigger CI

charmander commented on README.rst:
```diff
@@ -217,20 +217,29 @@ As of 3.0.0 the ``$2y$`` prefix is still supported in ``hashpw`` but deprecated.
     ...     bcrypt.gensalt()
     ... )
 
+Using a hash function without a hash is `recommended against`_ as it may expose
```
(without a salt)

FWDekker replied:
Feedback implemented in ee8a991fbb837f76c7711a00dc63fab293d00bfc. Let me know what you think :)
